arch/arm/core/userspace.S

```diff
@@ -0,0 +1,178 @@
+/*
+ * Userspace and service handler hooks
+ *
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+#include <offsets_short.h>
+#include <toolchain.h>
+#include <linker/sections.h>
+#include <kernel_structs.h>
+#include <arch/cpu.h>
+#include <syscall.h>
+
+_ASM_FILE_PROLOGUE
+
+GTEXT(_arm_userspace_enter)
+GTEXT(_arm_do_syscall)
+GDATA(_kernel)
+
+/* Imports */
+GTEXT(_k_syscall_table)
+
+/**
+ *
+ * User space entry function
+ *
+ * This function is the entry point to user mode from privileged execution.
+ * The conversion is one way, and threads which transition to user mode do
+ * not transition back later, unless they are doing system calls.
+ *
+ */
+SECTION_FUNC(TEXT,_arm_userspace_enter)
+    /* preserve arguments */
+    mov lr, r0
+
+    /*
+     * calculate end of stack and set SP to it
+     */
+    ldr r0, [sp]
+    ldr ip, [sp,#4]
+    add r0, ip
+
+    msr PSP, r0
+
+    /* push args to stack */
+    push {r1,r2,r3,lr}
+
+    /* clear the stack area to clean out privileged data */
+    /* from right past the guard right up to the args */
+    mrs r0, PSP
+#ifdef CONFIG_INIT_STACKS
+    ldr r1,=0xaaaaaaaa
+#else
+    ldr r1,=0
+#endif
+    mov r2, ip
+    sub r2, #16		/* don't erase the pushed regs */
+#ifdef CONFIG_MPU_STACK_GUARD
+    ldr r3, =MPU_GUARD_ALIGN_AND_SIZE
+    sub r2, r3
+    add r0, r3
+#endif
+    bl memset
+
+    /* setup arguments to thread_entry function */
+    ldr r0, =_kernel
+    ldr r0, [r0, #_kernel_offset_to_current]
+    bl configure_mpu_user_stack_context
+
+    pop {r0,r1,r2,lr}
+    /* change processor mode to unprivileged */
+    mov ip, #3
+    msr CONTROL, ip
+
+    /* jump to user entry */
+    bx lr
+
+/**
+ *
+ * Userspace system call function
+ *
+ * This function is used to do system calls from unprivileged code.  This
+ * function is responsible for the following:
+ * 1) Fixing up bad syscalls
+ * 2) Configuring privileged stack and loading up stack arguments
+ * 3) Dispatching the system call
+ * 4) Restoring stack and calling back to the caller of the SVC
+ *
+ */
+SECTION_FUNC(TEXT, _arm_do_syscall)
+    /*
+     * r0-r3 are values from pre-SVC from stack frame stored during SVC
+     * 16 bytes of storage reside on the stack:
+     * arg5, arg6, call_id, and LR from SVC frame
+     */
+    push {r4,r5,r6,lr}
+
+    ldr r4, =_SYSCALL_LIMIT
```

Comment:
we already validated the system call ID in _do_syscall, why are we doing it again here?

Reply:
because at the end of the svc, all of the registers are reset back to what they were before and the exception frame is popped off the stack.  So at the point where we enter the actual syscall, we have r0/r1/r2/r3 holding arguments, and the stack holding arg5/arg6/call_id/lr.  If I overwrite the call_id with the SYSCALL_BAD, then we lose the call_id information.

I suppose I could go ahead and overwrite the arg0 with the bogus sycall value, and overwrite the call_id stack entry with the SYSCALL_BAD and just dispatch it like usual.